flytekit/image_spec/image_spec.py

```diff
@@ -79,6 +79,22 @@ def __post_init__(self):
         if self.registry:
             self.registry = self.registry.lower()
 
+    @property
```

Comment:
We can use `functools.cached_property` to cache this: https://docs.python.org/3/library/functools.html#functools.cached_property.

Reply:
this is cool. I'll update it

Reply:
Tests fail after I change to `functools.cached_property`. Maybe it's because __hash__ also calls id(), which could cause a race condition

Reply:
added it